src/traces/image/plot.js

```diff
@@ -82,48 +95,118 @@ module.exports = function plot(gd, plotinfo, cdimage, imageLayer) {
         var image3 = plotGroup.selectAll('image')
-            .data(cd);
+            .data(data);
 
         image3.enter().append('svg:image').attr({
             xmlns: xmlnsNamespaces.svg,
             preserveAspectRatio: 'none'
         });
 
-        image3.attr({
-            height: imageHeight,
-            width: imageWidth,
-            x: left,
-            y: top,
-            'xlink:href': canvas.toDataURL('image/png')
+        if(fastImage) sizeImage(image3);
+        image3.exit().remove();
+
+        // Pixelated image rendering
+        // http://phrogz.net/tmp/canvas_image_zoom.html
+        // https://developer.mozilla.org/en-US/docs/Web/CSS/image-rendering
+        image3
+            .attr('style', 'image-rendering: optimizeSpeed; image-rendering: -moz-crisp-edges; image-rendering: -o-crisp-edges; image-rendering: -webkit-optimize-contrast; image-rendering: optimize-contrast; image-rendering: crisp-edges; image-rendering: pixelated;');
+
+        var p = new Promise(function(resolve) {
+            if(trace._isFromZ) {
+                resolve();
+            } else if(trace._isFromSource) {
+                // Transfer image to a canvas to access pixel information
+                trace._canvas = trace._canvas || document.createElement('canvas');
+                trace._canvas.width = w;
+                trace._canvas.height = h;
+                var context = trace._canvas.getContext('2d');
+
+                var sel;
+                if(fastImage) {
+                    // Use the displayed image
+                    sel = image3;
+                } else {
+                    // Use the hidden image
+                    sel = d3.select(image3[0][1]);
+                }
+
+                var image = sel.node();
+                image.onload = function() {
+                    context.drawImage(image, 0, 0);
+                    resolve();
+                };
+                sel.attr('xlink:href', trace.source);
+            }
+        })
+        .then(function() {
+            if(!fastImage) {
+                var canvas;
+                if(trace._isFromZ) {
+                    canvas = drawMagnifiedPixelsOnCanvas(function(i, j) {return z[j][i];});
+                } else if(trace._isFromSource) {
+                    var context = trace._canvas.getContext('2d');
+                    var data = context.getImageData(0, 0, w, h).data;
+                    canvas = drawMagnifiedPixelsOnCanvas(function(i, j) {
+                        var index = 4 * (j * w + i);
+                        return [
+                            data[index + 0],
+                            data[index + 1],
+                            data[index + 2],
+                            data[index + 3]
+                        ];
+                    });
+                }
+                var href = canvas.toDataURL('image/png');
+                var displayImage = d3.select(image3[0][0]);
```

Comment:
Have you tested this when multiple images are present using source?

Reply:
Wait, is `d3.select(image3[0][0])` different in a useful way from just `image3`?
If it is though, do `image3.node()` instead of `image3[0][0]`.

Reply:
> Wait, is `d3.select(image3[0][0])` different in a useful way from just `image3`?

At this point in the code, we want to update and style the displayed image but not the hidden one which is also included in the `image3` selection.

> If it is though, do `image3.node()` instead of `image3[0][0]`.

Done in 09eb1d0c144692a2fe4590e524651bd7e8cb6215